internal/confmapprovider/configsource/provider.go

```diff
@@ -0,0 +1,165 @@
+// Copyright Splunk, Inc.
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configsource
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.opentelemetry.io/collector/confmap"
+	"go.uber.org/zap"
+
+	"github.com/signalfx/splunk-otel-collector/internal/configsource"
+	"github.com/signalfx/splunk-otel-collector/internal/configsource/envvarconfigsource"
+	"github.com/signalfx/splunk-otel-collector/internal/configsource/etcd2configsource"
+	"github.com/signalfx/splunk-otel-collector/internal/configsource/includeconfigsource"
+	"github.com/signalfx/splunk-otel-collector/internal/configsource/vaultconfigsource"
+	"github.com/signalfx/splunk-otel-collector/internal/configsource/zookeeperconfigsource"
+)
+
+var configSourceFactories = func() configsource.Factories {
+	factories := make(configsource.Factories)
+	for _, f := range []configsource.Factory{
+		envvarconfigsource.NewFactory(),
+		includeconfigsource.NewFactory(),
+		vaultconfigsource.NewFactory(),
+		zookeeperconfigsource.NewFactory(),
+		etcd2configsource.NewFactory(),
+	} {
+		if _, ok := factories[f.Type()]; ok {
+			panic(fmt.Sprintf("duplicate config source factory %q", f.Type()))
+		}
+		factories[f.Type()] = f
+	}
+	return factories
+}()
+
+type Hook interface {
+	OnNew()
+	OnRetrieve(scheme string, retrieved map[string]any)
+	OnShutdown()
+}
+
+type Provider interface {
```

Comment:
Can we make this easier to understand for reviewers and maintainers? Looks like this interface is called Provider, its method is called Provider, and the method accepts a Provider and returns a Provider :). Do we need the interface -- it seems like the only implementer is ProviderWrapper (which is exported). And can we rename the `Provider` function to `WrapProvider` (or something) if that's what it does?

Reply:
It's a `configsource.Provider` that takes a `confmap.Provider` and returns a `confmap.Provider`. I initially named the method `Wrap` but thought there were too many "wrap" prefixes so relied on the package name to help differentiate. The interface is helpful imo to help keep the surface area minimal while allowing changes to the implementation. Went ahead and unexported the implementation since the `confmap.Provider` is what's important. Also updated to use the clearer method name and added some doc comments.